File: Hackerrank/30 Days of Code/Day12_Inheritance.cpp
#include <iostream>
#include <vector>

using namespace std;


class Person{
	protected:
		string firstName;
		string lastName;
		int id;
	public:
		Person(string firstName, string lastName, int identification){
			this->firstName = firstName;
			this->lastName = lastName;
			this->id = identification;
		}
		void printPerson(){
			cout<< "Name: "<< lastName << ", "<< firstName <<"\nID: "<< id << "\n"; 
		}
	
};

class Student :  public Person{
	private:
		vector<int> testScores;  
        int scoreNum =0;
        double avg =0.0;
        int scoreSum=0;
	public:
        /*	
        *   Class Constructor
        *   
        *   Parameters:
        *   firstName - A string denoting the Person's first name.
        *   lastName - A string denoting the Person's last name.
        *   id - An integer denoting the Person's ID number.
        *   scores - An array of integers denoting the Person's test scores.
        */
        // Write your constructor here
        Student(string firstName, string lastName, int identification, vector<int> testScores)
            :Person(firstName, lastName, identification){
			this->testScores = testScores;
            this->scoreNum=testScores.size();
		}
                    

        /*	
        *   Function Name: calculate
        *   Return: A character denoting the grade.
        */
        // Write your function here
        char calculate(){
            char grade=0;
            
            for(int i; i<scoreNum; i++){
                scoreSum = scoreSum + testScores[i];
            }
            
            avg = scoreSum / scoreNum;
            
            if(avg >= 90 && avg<=100)
                grade='O';
            else if(avg >= 80 && avg<90)
                grade='E';
            else if(avg >= 70 && avg<80)
                grade='A';
            else if(avg >= 55 && avg<70)
                grade='P';
            else if(avg >= 40 && avg<55)
                grade='D';
            else if(avg >= 0 && avg<40)
                grade='T';
            else
                grade='T';      

            
            return grade;
        }
};

int main() {
	string firstName;
  	string lastName;
	int id;
  	int numScores;
	cin >> firstName >> lastName >> id >> numScores;
  	vector<int> scores;
  	for(int i = 0; i < numScores; i++){
	  	int tmpScore;
	  	cin >> tmpScore;
		scores.push_back(tmpScore);
	}
	Student* s = new Student(firstName, lastName, id, scores);
	s->printPerson();
	cout << "Grade: " << s->calculate() << "\n";
	return 0;
}

File: Hackerrank/30 Days of Code/Day12_Inheritance.py
class Person:
	def __init__(self, firstName, lastName, idNumber):
		self.firstName = firstName
		self.lastName = lastName
		self.idNumber = idNumber
	def printPerson(self):
		print("Name:", self.lastName + ",", self.firstName)
		print("ID:", self.idNumber)
		
class Student(Person):
    #   Class Constructor
    #   
    #   Parameters:
    #   firstName - A string denoting the Person's first name.
    #   lastName - A string denoting the Person's last name.
    #   id - An integer denoting the Person's ID number.
    #   scores - An array of integers denoting the Person's test scores.
    #
    # Write your constructor here
    def __init__(self, firstName, lastName, id, scores):
        super().__init__(firstName, lastName, id)
        self.scores = scores
    

    #   Function Name: calculate
    #   Return: A character denoting the grade.
    #
    # Write your function here
    def calculate(self):
        scoreNum = len(self.scores)
        scoreSum=0
        avg=0.0
        grade=""
        for i in range(scoreNum):
            scoreSum += scores[i]
            
        avg = float(scoreSum) / scoreNum
        
        if(avg >= 90 and avg<=100):
            grade='O'
        elif(avg >= 80 and avg<90):
            grade='E'
        elif(avg >= 70 and avg<80):
            grade='A'
        elif(avg >= 55 and avg<70):
            grade='P'
        elif(avg >= 40 and avg<55):
            grade='D'
        elif(avg >= 0 and avg<40):
            grade='T'
        else:
            grade='T'
            
        return grade

		
line = input().split()
firstName = line[0]
lastName = line[1]
idNum = line[2]
numScores = int(input()) # not needed for Python
scores = list( map(int, input().split()) )
s = Student(firstName, lastName, idNum, scores)
s.printPerson()
print("Grade:", s.calculate())